File: ruby/lib/xi/ml/classify/prediction_statistics.rb
# encoding: utf-8



# Class displaying and storing the statistics of given prediction
class Xi::ML::Classify::PredictionStatistics < Xi::ML::Tools::Component
  attr_reader :categories, :data_files, :stats, \
    :confusion_matrix, :real_categories, :predicted_categories

  # Initialize
  #
  # @param data_files [Array] list of data files with classified documents
  # @param categories [Array] list of document categories
  def initialize(data_files, categories)
    super()

    @stats = {}
    @data_files = data_files.dup
    @categories = categories.dup
  end

  # Compute the accuracy, precision and recall statistics;
  # Save statistics to json file
  #
  # @param output [String] the file where to store statistics
  def save_stats(output)
    @logger.info("Save statistics to '#{output}' file")

    compute_confusion_matrix()
    compute_accuracy_stats()
    compute_pr_stats()

    Xi::ML::Tools::Utils.create_path(output)
    File.open(output, 'w') {|stream| stream.puts(JSON.pretty_generate(@stats)) }
  end

  # Compute the confusion matrix
  def compute_confusion_matrix
    @cm = {}
    @cm['total'] = Hash.new(0)

    @real_categories = []
    @predicted_categories = []

    @data_files.each do |input_file|
      sc = Xi::ML::Corpus::StreamCorpus.new(input_file)

      sc.each_doc do |doc|
        if doc['category'] && doc['season']
          real_category = doc['category']
          predicted_category = doc['season']

          @real_categories << real_category \
            unless @real_categories.include?(real_category)

          @predicted_categories << predicted_category \
            unless @predicted_categories.include?(predicted_category)

          @cm[real_category] = Hash.new(0) unless @cm.key?(real_category)
          @cm[real_category][predicted_category] += 1
          @cm[real_category]['total'] += 1

          @cm['total'][predicted_category] += 1
          @cm['total']['total'] += 1
        end
      end
    end

    @real_categories.sort!
    @predicted_categories.sort!

    display_confusion_matrix()
  end

  def display_confusion_matrix
    rcategories = @real_categories.dup << 'total'
    pcategories = @predicted_categories.dup << 'total'

    n = (rcategories | pcategories).map{|x| x.size }.max + 3

    display = ' ' * n
    pcategories.each{|pcat| display << pcat.rjust(n) }

    display << "\n"
    rcategories.each do |rcat|
      display << rcat.ljust(n)
      pcategories.each{|pcat| display << @cm[rcat][pcat].to_s.rjust(n) }
      display << "\n"
    end

    @logger.info("Confusion matrix and marginals\n#{display}")
  end

  # Compute the accuracy stats
  def compute_accuracy_stats
    gc, gt = 0, 0

    # print stats by class
    @real_categories.each do |cat|
      correct = @cm[cat][cat] ? @cm[cat][cat] : 0
      total = @cm[cat]['total'] ? @cm[cat]['total'] : 0
      avg_acc = percentage(correct, total)

      @logger.info("Correctly classified '#{cat}' documents : "\
        "#{correct} / #{total} = #{avg_acc} %")

      # global counts
      gc += correct
      gt += total
    end

    # global accuracy
    g_acc = percentage(gc, gt)
    @logger.info("Correctly classified documents: #{gc} / #{gt} = #{g_acc} %")

    # store 'global accuracy' stat
    @stats[:global_accuracy] = g_acc
  end

  # Compute precision and recall statistics
  def compute_pr_stats
    # recall & precision stats for each class
    precision, recall = {}, {}

    @real_categories.each do |cat|
      correct = @cm[cat][cat] ? @cm[cat][cat] : 0
      ptotal = @cm['total'][cat] ? @cm['total'][cat] : 0
      rtotal = @cm[cat]['total'] ? @cm[cat]['total'] : 0

      precision[cat] = percentage(correct, ptotal)
      recall[cat] = percentage(correct, rtotal)

      # store P - R stats
      @stats[cat] = {
        precision: precision[cat],
        recall: recall[cat],
      }
    end

    # display the 1-versus-rest confusion matrix for each class
    @real_categories.each do |cat|
      tp = @cm[cat][cat] ? @cm[cat][cat] : 0
      fn = @cm[cat]['total'] - tp
      fp = @cm['total'][cat] - tp
      tn = @cm['total']['total'] - tp - fn - fp

      cmatrix = '=' * 40 << "\n"
      cmatrix << "Class=#{cat}\n"
      cmatrix << '=' * 40 << "\n"
      cmatrix << "      | declare H1 |  declare H0 |\n"
      cmatrix << "is H1 | #{tp.to_s.rjust(10)} | #{fn.to_s.rjust(11)} |\n"
      cmatrix << "is H0 | #{fp.to_s.rjust(10)} | #{tn.to_s.rjust(11)} |\n"
      cmatrix << '-' * 40 << "\n"
      cmatrix << "Precision = #{precision[cat]} %\n"
      cmatrix << "Recall    = #{recall[cat]} %\n"

      @logger.info("Confusion matix, precision, recall for #{cat}\n#{cmatrix}")
    end
  end

  # Own percentage method to account for zero division error
  def percentage(x, y)
    if y == 0
      @logger.warn('Zero division error')
      return 0.0
    end
    ((x.to_f / y) * 100).round(2)
  end

  private :compute_pr_stats, :compute_accuracy_stats, :compute_pr_stats, \
    :percentage
end

File: python/lib/xi/ml/classify/prediction_statistics.py
# -*-coding:utf-8 -*


import json

from xi.ml.common import Component
from xi.ml.tools import utils
from xi.ml.corpus import StreamCorpus


class PredictionStatistics(Component):
    """Class displaying and storing the statistics of given prediction"""

    def __init__(self, data_files, categories):
        """Load classified data files and list of categories"""

        super().__init__()

        for filename in data_files:
            utils.check_file_readable(filename)

        self.stats = {}
        self.data_files = list(data_files)
        self.categories = list(categories)

    def compute_stats(self):
        """Compute the accuracy, precision and recall statistics"""

        self.stats = {}

        tp, fn, fp, tn = {}, {}, {}, {}
        n_total, n_incorrect = {}, {}

        # init stats dictionary
        # - number of total documents in class 'category'
        # - number of correctly classified documents in class 'category'
        for category in self.categories:
            tp[category] = 0
            fn[category] = 0
            fp[category] = 0
            tn[category] = 0
            n_total[category] = 0
            n_incorrect[category] = 0

        # count correct predictions in given data files
        for input_file in self.data_files:
            sc = StreamCorpus(input_file)

            for doc in sc:
                if 'category' not in doc or 'season' not in doc:
                    continue

                real_categories = doc['category']
                predicted_categories = doc['season']

                # convert real_category to list
                if isinstance(real_categories, str):
                    real_categories = [real_categories]

                # convert predicted_category to list
                if isinstance(predicted_categories, str):
                    predicted_categories = [predicted_categories]

                for real_category in real_categories:
                    n_total[real_category] += 1

                    if real_category in predicted_categories:
                        tp[real_category] += 1
                    else:
                        for predicted_category in predicted_categories:
                            fp[predicted_category] += 1

        if sum(n_total.values()) == 0:
            self.logger.warning('No data to process')
            return

        # print stats by class
        gc, gt = 0, 0
        for category in self.categories:
            total = n_total[category]
            correct = tp[category]
            avg_accuracy = self.div(correct, total)

            self.logger.info(
                "Correctly classified documents of class='{}': "
                "{} / {} = {:.2%}"
                .format(category, correct, total, avg_accuracy))

            # global counts
            gc += correct
            gt += total

        # global accuracy
        global_accuracy = self.div(gc, gt)

        # print global stats
        self.logger.info(
            "Correctly classified documents: {} / {} = {:.2%}"
            .format(gc, gt, global_accuracy))

        # recall & precision stats for each class
        # true-positive, false-negative, false-positive, true-negative ratios
        recall, precision = {}, {}
        for category in self.categories:
            fn[category] = n_total[category] - tp[category]
            tn[category] = gt - tp[category] - fn[category] - fp[category]

            precision[category] = self.div(
                tp[category], tp[category] + fp[category])

            recall[category] = self.div(
                tp[category], tp[category] + fn[category])

        # print the confusion matrix for each class
        for category in self.categories:
            table = "=" * 40 + "\n"
            table += "Class={}\n".format(category)
            table += "=" * 40 + "\n"
            table += "      | declare H1 |  declare H0 |\n"

            table += "is H1 | {:>10} | {:>11} |\n"\
                .format(tp[category], fn[category])

            table += "is H0 | {:>10} | {:>11} |\n".format(
                fp[category], tn[category])

            table += "-" * 40 + "\n"

            table += "Precision = {:.2%}\n".format(
                precision[category])

            table += "Recall    = {:.2%}\n".format(
                recall[category])

            self.logger.info(
                "Confusion matix, precision and recall stats "
                "for the '{}' class:\n{}".format(category, table))

        # store stats
        self.stats['global-accuracy'] = "{:.2%}".format(global_accuracy)
        for category in self.categories:
            self.stats[category] = {}
            self.stats[category]['precision'] = "{:.2%}".format(
                precision[category])
            self.stats[category]['recall'] = "{:.2%}".format(
                recall[category])

    def store_stats(self, output):
        """Store statistics on predictions"""

        self.logger.info(
            "Save statistics on predictions to '{}'"
            .format(output))

        utils.create_path(output)
        with open(output, 'w') as ostream:
            ostream.write(json.dumps(self.stats, indent=2))

    def div(self, x, y):
        """Own div method to account for zero division error"""

        if y != 0:
            return x / y
        else:
            self.logger.warning('Zero division error')
            return 0
